src/instrument.ts

```diff
@@ -0,0 +1,262 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {
+  SEMATTRS_DB_NAME,
+  SEMATTRS_DB_STATEMENT,
+  SEMATTRS_DB_SYSTEM,
+  SEMATTRS_DB_SQL_TABLE,
+} from '@opentelemetry/semantic-conventions';
+
+const {TracerProvider} = require('@opentelemetry/sdk-trace-node');
+
+import {
+  ContextManager,
+  Span,
+  SpanStatusCode,
+  context,
+  trace,
+  INVALID_SPAN_CONTEXT,
+  SpanAttributes,
+  TimeInput,
+  Link,
+  Exception,
+  SpanContext,
+  SpanStatus,
+  SpanKind,
+} from '@opentelemetry/api';
+
+let optedInPII: boolean =
+  process.env.SPANNER_ENABLE_EXTENDED_TRACING === 'true';
+
+interface SQLStatement {
+  sql: string;
+}
+
+interface observabilityOptions {
+  tracerProvider: typeof TracerProvider;
+  enableExtendedTracing: boolean;
+}
+
+export type {observabilityOptions as ObservabilityOptions};
+
+export function applyObservabilityOptions(
+  opts: observabilityOptions | undefined
+) {
+  if (!opts) {
+    return;
+  }
+
+  if (opts.tracerProvider) {
+    setTracerProvider(opts.tracerProvider);
+  }
+
+  if (opts.enableExtendedTracing) {
+    optInForSQLStatementOnSpans();
+  }
+}
+
+let defaultTracerProvider: typeof TracerProvider = undefined;
+
+// setTracerProvider allows the caller to hook up an OpenTelemetry
+// TracerProvider that spans generated from this library shall be attached to,
+// instead of using the global configuration. Later on if `getTracer` is invoked and
+// the default tracerProvider is unset, it'll use the global tracer
+// otherwise it'll use the set TracerProvider.
+export function setTracerProvider(freshTracerProvider: typeof TracerProvider) {
+  defaultTracerProvider = freshTracerProvider;
+}
+
+const TRACER_NAME = 'nodejs-spanner';
+
+// getTracer fetches the tracer each time that it is invoked to solve
+// the problem of observability being initialized after Spanner objects
+// have been already created.
+export function getTracer(config?: traceConfig) {
```

Comment:
If observabilityOptions will be injected via SpannerOptions, in that case won't all the classes will have access to these options, in what scenario this will be used ?

Reply:
Use observabilityOptions instead of traceConfig

Reply:
I just passed in TracerProvider directly. Thank you!